sdk/ai/azure-ai-generative/azure/ai/generative/evaluate/_metrics_handler/_code_metric_handler.py

```diff
@@ -0,0 +1,108 @@
+            ground_truth_column_name=None,
+            metrics_mapping=None,
+            metrics=None,
+            type_to_kwargs=None,
+    ):
+
+        super().__init__(
+            task_type=task_type,
+            prediction_data=prediction_data,
+            test_data=test_data,
+            truth_data=truth_data,
+            prediction_data_column_name=prediction_data_column_name,
+            ground_truth_column_name=ground_truth_column_name,
+            metrics_mapping=metrics_mapping,
+            metrics=metrics,
+            type_to_kwargs=type_to_kwargs,
+        )
+
+        self._validate()
+
+    def _validate(self):
+        supported_list = [isinstance(metric, CodeMetric) for metric in self.metrics]
+        if not all(supported_list):
+            raise Exception(f"{self.__class__.__name__} supports only {CodeMetric.__class__.__name__} type of metrics")
+
+    def calculate_metrics(self):
+        LOGGER.info(f"Calculating code metrics : {[metric.name for metric in self.metrics]}")
+        metrics_dict = {"artifacts": {}, "metrics": {}}
+        metric_results_futures = {}
+        with tqdm.tqdm(total=len(self.metrics)) as progress_bar:
+            with ThreadPoolExecutor(thread_name_prefix="code_metrics") as thread_pool:
+                for metric in self.metrics:
+                    metric_values = []
+                    metric_results_futures.update({metric.name: thread_pool.submit(
+                        self._calculate_metric, metric, self.test_data.to_dict('records'),
+                        self.prediction_data.to_dict('records') if self.prediction_data is not None else None
+                    )})
+
+                for metric_name, metric_result_future in metric_results_futures.items():
+                    try:
+                        metric_result = metric_result_future.result()
+                        metrics_dict["artifacts"].update(metric_result["artifacts"])
+                        if "metrics" in metric_result.keys() and metric_result["metrics"] is not None:
+                            metrics_dict["metrics"].update(metric_result["metrics"])
+                        progress_bar.update(1)
+                    except Exception as ex:
+                        progress_bar.update(1)
+                        LOGGER.info(f"Error calculating value for {metric_name}, failed with error {str(ex)} : Stack trace : {str(ex.__traceback__)}")
+
+        return metrics_dict
+
+    def _submit_method(self, method, *args, **kwargs):
+        import inspect
+        if inspect.iscoroutinefunction(method):
+            import asyncio
+            return asyncio.run(method(*args, **kwargs))
+        return method(*args, **kwargs)
+
+    def _calculate_metric(self, metric, data, response):
+        row_metric_futures = []
+        row_metric_result = []
+        aggregated_metrics = None
+
+        with ThreadPoolExecutor(thread_name_prefix="code_metrics_row") as thread_pool:
```

Comment:
Wow, a thread pool inside a thread pool. This logic will go away after PF integration hopefully.

Reply:
Yes it will. This is to parallelize at data and at metric level